content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,388 @@
+
+#### Alternative 2: Fully uncontrolled component
+
+Another alternative would be for our component to fully own the "draft" email state. This might be helpful if the "committed" state lives in a state container like Redux. In that case, our component could still accept a prop for the _initial_ value, but it would ignore any changes to that prop afterward. For example:
+
+```js
+class EmailInput extends Component {
+  state = {
+    email: this.props.defaultEmail
+  };
+
+  handleChange = event => {
+    this.setState({ email: event.target.value });
+  };
+
+  render() {
+    return <input onChange={this.handleChange} value={this.state.email} />;
+  }
+}
+```
+
+This design makes the component a little easier to use, but it also has a downside. Let's say that our "edit profile" form could be used to edit multiple users. Whenever a new user is selected, the form should reset the email field. If our `EmailInput` is uncontrolled, it will ignore any changes to `props.defaultEmail`. So how could we support this use case?
+
+#### Option 1: Reset uncontrolled component with a `key`
+
+One approach would be to use a special React attribute called `key`. React uses this property to decide whether to [_update_ a component instance or _create_ a new one](/docs/reconciliation.html#keys). In our case, we could use an attribute like a user id to recreate the email input any time a new user is selected. Each time the `EmailInput` is recreated, it will "reset" its state to the value specified by the `defaultEmail` prop.
+
+```js
+<EmailInput
+  defaultEmail={this.props.user.email}
+  key={this.props.user.id}
+/>
+```
+
+> Note
+>
+> With this approach, you don't have to add `key` to every input. Instead, it might make sense to put a `key` on the whole form. Every time the key changes, all components within the form will be re-created with a freshly initialized state. While this may sound slow, in practice the difference is often insignificant. It can even be faster if the components have heavy logic that runs on updates.
+
+#### Option 2: Reset uncontrolled component with an id prop
+
+Recreating this simple example component is cheap, but what if the component was expensive to initialize? Or what if it had other state that we _did not want to reset_?
+
+A slight variation on the above approach would be to watch for changes in a prop like "id" and use it to reset state:
+
+```js
+class EmailInput extends Component {
+  state = {
+    email: this.props.defaultEmail,
+    prevPropsID: this.props.id
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    // Any time the current user changes,
+    // Reset any parts of state that are tied to that user.
+    // In this simple example, that's just the email.
+    if (props.id !== state.prevPropsID) {
```

Comment:
I can imagine somebody thinking `id` is a special thing like `key`. Maybe let's make it `userID` as a more obviously custom one?